Src/PerformanceCollector/Shared/Implementation/QuickPulse/QuickPulseServiceClient.cs

```diff
@@ -0,0 +1,228 @@
+            this.timeout = timeout ?? this.timeout;
+        }
+
+        public Uri ServiceUri { get; }
+
+        public bool? Ping(string instrumentationKey, DateTime timestamp)
+        {
+            MemoryStream bodyStream = this.WritePingData(timestamp);
+
+            var path = string.Format(CultureInfo.InvariantCulture, "ping?ikey={0}", instrumentationKey);
+            HttpWebResponse response = this.SendRequest(WebRequestMethods.Http.Post, path, bodyStream, true);
+            if (response == null)
+            {
+                return null;
+            }
+            
+            return ProcessResponse(response);
+        }
+        
+        public bool? SubmitSamples(IEnumerable<QuickPulseDataSample> samples, string instrumentationKey)
+        {
+            MemoryStream bodyStream = WriteSamples(samples, instrumentationKey);
+            
+            var path = string.Format(CultureInfo.InvariantCulture, "post?ikey={0}", HttpUtility.UrlEncode(instrumentationKey));
+            HttpWebResponse response = this.SendRequest(WebRequestMethods.Http.Post, path, bodyStream, false);
+            if (response == null)
+            {
+                return null;
+            }
+
+            return ProcessResponse(response);
+        }
+
+        private static bool? ProcessResponse(HttpWebResponse response)
+        {
+            bool isSubscribed;
+            if (!bool.TryParse(response.GetResponseHeader(XMsQpsSubscribedHeaderName), out isSubscribed))
+            {
+                return null;
+            }
+
+            return isSubscribed;
+        }
+
+        private static double Round(double value)
+        {
+            return Math.Round(value, 4, MidpointRounding.AwayFromZero);
+        }
+
+        private MemoryStream WritePingData(DateTime timestamp)
+        {
+            var bodyStream = new MemoryStream();
+
+            var dataPoint = new MonitoringDataPoint
+            {
+                Version = this.version,
+                //InstrumentationKey = instrumentationKey,
+                Instance = this.instanceName,
+                Timestamp = timestamp
+            };
+
+            var serializer = new DataContractJsonSerializer(typeof(MonitoringDataPoint));
+            serializer.WriteObject(bodyStream, dataPoint);
+            bodyStream.Position = 0;
+
+            return bodyStream;
+        }
+
+        private MemoryStream WriteSamples(IEnumerable<QuickPulseDataSample> samples, string instrumentationKey)
+        {
+            var bodyStream = new MemoryStream();
+            var monitoringPoints = new List<MonitoringDataPoint>();
+
+            foreach (var sample in samples)
+            {
+                var metricPoints = new List<MetricPoint>
+                                       {
+                                           new MetricPoint
+                                               {
+                                                   Name = @"\ApplicationInsights\Requests/Sec",
```

Comment:
These are in one place only. Looks kind of ugly, but I really see no value in moving them out, they have to be somewhere.